collector/runtime-benchmarks/bufreader/src/main.rs

```diff
@@ -1,36 +1,37 @@
-use std::io::{BufRead, BufReader, Write};
+use std::io::{BufRead, BufReader, Cursor, Write};
 
 use snap::{read::FrameDecoder, write::FrameEncoder};
 
-use benchlib::benchmark::{black_box, run_benchmark_group};
-use benchlib::define_benchmark;
+use benchlib::{black_box, define_benchmark, run_benchmark_group};
 
-const BYTES: usize = 64 * 1024 * 1024;
+// Inspired by https://github.com/rust-lang/rust/issues/102727
+// The pattern we want is a BufReader which wraps a Read impl where one Read::read call will
+// never fill the whole BufReader buffer.
+fn read_buf(
+    mut reader: BufReader<FrameDecoder<Cursor<Vec<u8>>>>,
+) -> BufReader<FrameDecoder<Cursor<Vec<u8>>>> {
+    while let Ok(buf) = reader.fill_buf() {
+        if buf.is_empty() {
+            break;
+        }
+        black_box(buf);
+        let len = buf.len();
+        reader.consume(len);
+    }
+    reader
+}
 
 fn main() {
-    // Inspired by https://github.com/rust-lang/rust/issues/102727
-    // The pattern we want is a BufReader which wraps a Read impl where one Read::read call will
-    // never fill the whole BufReader buffer.
     run_benchmark_group(|group| {
-        define_benchmark!(group, bufreader_snappy, {
-            let data = vec![0u8; BYTES];
-            move || {
-                let mut compressed = Vec::new();
-                FrameEncoder::new(&mut compressed)
-                    .write_all(&data[..])
-                    .unwrap();
-                let mut reader =
-                    BufReader::with_capacity(BYTES, FrameDecoder::new(&compressed[..]));
+        define_benchmark!(group, "bufreader_snappy", read_buf, {
```

Comment:
This is a complex `$init` value. Is its execution measured as part of the benchmark?

Reply:
It shouldn't be, and that is also the point of why the constructor exists. So that I can call the benchmark in a two part way:
1) Create init data
2) Start measuring
3) Pass init data and execute benchmark
4) (stop measuring)

Reply:
That all sounds reasonable, it's just not currently clear. More documentation please! :)